File: pong/ball.py
"""
ball.py -- Unified Ball class for both Classic and Physics modes
"""

import pygame
import numpy as np
import math
from pong.physics_object import PhysicsObject
from pong.constants import *

class Ball(PhysicsObject):
    """
    A unified ball class supporting both classic and physics-based Pong modes.

    In 'physics' mode: Supports spin (Magnus effect), visual trail, and fire effects.
    In 'classic' mode: Simple ball with basic movement.
    """

    def __init__(self, x, y, radius, color, mass=1, vel=(0, 0), mode='physics'):
        """
        Initializes the ball with position, size, color, and optional velocity.

        Args:
            x (float): Initial X position.
            y (float): Initial Y position.
            radius (int): Ball radius in pixels.
            color (tuple): RGB color value.
            mass (float): Ball mass.
            vel (tuple): Initial velocity vector (vx, vy).
            mode (str): 'classic' or 'physics'
        """
        super().__init__(pos=(x, y), mass=mass, vel=vel)
        self.radius = radius
        self.color = color
        self.spin = 0
        self.mode = mode

        self.trail = []
        self.max_trail = 10
        self.original_pos = np.array([x, y], dtype=float)
        self.original_vel = np.array(vel, dtype=float)

    def draw(self, win):
        """
        Draws the ball on the game window.
        Uses simple circle for classic mode, fire effects for physics mode.
        """
        if self.mode == 'classic':
            self._draw_classic(win)
        else:
            self._draw_physics(win)

    def _draw_classic(self, win):
        """Simple classic mode drawing - just a circle"""
        pygame.draw.circle(win, self.color, (int(self.pos[0]), int(self.pos[1])), self.radius)

    def _draw_physics(self, win):
        """Physics mode drawing with fire trail, aura, and spin effects"""
        # Fire colors for the trail (from dark to bright)
        FIRE_COLORS = [DARK_RED, SCARLET, ORANGE_RED, ORANGE, GOLD, YELLOW, WHITE]
        ALL_FIRE_COLORS = [DARK_RED, SCARLET, ORANGE_RED, ORANGE, GOLD, YELLOW, WHITE, GOLD, YELLOW, WHITE]
        color_cycle_len = len(ALL_FIRE_COLORS)

        speed = self.speed
        max_speed = 30
        acc_magnitude = np.linalg.norm(self.acc)
        t = min(acc_magnitude / 15, 1)
        color_idx = int(t * (len(FIRE_COLORS) - 1))
        core_color = FIRE_COLORS[color_idx]

        # --- 1. Draw fire trail ---
        for i, pos in enumerate(self.trail):
            trail_t = i / max(len(self.trail)-1, 1)
            flicker = np.random.randint(-15, 16) if speed > 7 else 0
            alpha = int(max(35, 140 - i * 18 + flicker))
            radius = max(2, int(self.radius * (1 - trail_t * 0.8)))

            cidx = int((color_idx + i) % color_cycle_len)
            trail_color = ALL_FIRE_COLORS[cidx]
            surface = pygame.Surface((radius * 2, radius * 2), pygame.SRCALPHA)
            pygame.draw.circle(surface, (*trail_color, alpha), (radius, radius), radius)
            win.blit(surface, (pos[0] - radius, pos[1] - radius))

        # --- 2. Glowing aura if fast ---
        if speed > 6:
            for layer in range(3, 0, -1):
                aura_r = int(self.radius * (1.2 + 0.25 * layer))
                aura_alpha = 25 * layer + int(min(50, speed * 2))
                aura_color = FIRE_COLORS[min(len(FIRE_COLORS) - 1, color_idx + layer)]
                surf = pygame.Surface((aura_r * 2, aura_r * 2), pygame.SRCALPHA)
                pygame.draw.circle(surf, (*aura_color, aura_alpha), (aura_r, aura_r), aura_r)
                win.blit(surf, (int(self.pos[0]) - aura_r, int(self.pos[1]) - aura_r))

        # --- 3. Spin arc effect ---
        if abs(self.spin) > 4:
            direction = 0 if self.spin > 0 else math.pi
            arc_rect = pygame.Rect(int(self.pos[0] - self.radius), int(self.pos[1] - self.radius),
                                   self.radius * 2, self.radius * 2)
            arc_len = math.pi + min(math.pi * 0.8, (speed / max_speed) * math.pi * 0.8)
            base_thick = int(5 + min(18, (speed / max_speed) * 22))

            for j in range(5):
                arc_alpha = 60 + int((np.sin(pygame.time.get_ticks() / 40 + j) + 1) * 80) if speed > 8 else 50 + j*10
                arc_color = ALL_FIRE_COLORS[(color_idx + j) % color_cycle_len]
                thick = base_thick + j * 2
                arc_surf = pygame.Surface((self.radius * 2, self.radius * 2), pygame.SRCALPHA)
                pygame.draw.arc(arc_surf, (*arc_color, min(255, arc_alpha)),
                                pygame.Rect(0, 0, self.radius * 2, self.radius * 2),
                                direction, direction + arc_len, thick)
                win.blit(arc_surf, (int(self.pos[0]) - self.radius, int(self.pos[1]) - self.radius),
                         special_flags=pygame.BLEND_ADD)

        # --- 4. Main ball (core) ---
        if speed > 21:
            for k in range(2):
                pygame.draw.circle(win, WHITE if k else core_color,
                                   (int(self.pos[0]), int(self.pos[1])), self.radius - k, 0)
        else:
            pygame.draw.circle(win, core_color, (int(self.pos[0]), int(self.pos[1])), self.radius)

    def move(self):
        """Updates velocity and position. Physics mode applies Magnus effect."""
        if self.mode == 'physics':
            self.vel[1] += self.spin * 0.1  # Magnus effect
        self.pos += self.vel

    def update(self):
        """Updates the ball's state each frame."""
        if self.mode == 'physics':
            # Trail only for physics mode
            self.trail.insert(0, (int(self.pos[0]), int(self.pos[1])))
            if len(self.trail) > self.max_trail:
                self.trail.pop()
        self.move()

    def bounce_box(self, width, height):
        """Bounce off walls - used in menu and classic mode.
        Returns True if any bounce occurred, False otherwise."""
        self.move()
        bounced = False
        if self.pos[0] <= 0 or self.pos[0] >= width:
            self.vel *= [-1, 1]
            bounced = True
        if self.pos[1] <= 0 or self.pos[1] >= height:
            self.vel *= [1, -1]
            bounced = True
        return bounced

    def reset(self):
        """Resets ball to original position and velocity."""
        self.pos = self.original_pos.copy()
        if self.mode == 'classic':
            self.vel = np.array([-self.original_vel[0], 0], dtype=float)
        else:
            self.vel = -self.original_vel.copy()
        self.spin = 0
        self.trail.clear()

    @property
    def speed(self):
        """Returns the current speed (magnitude of velocity)."""
        return np.linalg.norm(self.vel)


# ----------------------- Legacy Alias for Backwards Compatibility -----------------------
class BallClassic(Ball):
    """
    @deprecated: Use Ball with mode='classic' instead
    """
    def __init__(self, x, y, radius, color, vel_x, vel_y):
        super().__init__(x, y, radius, color, mass=1, vel=(vel_x, vel_y), mode='classic')

File: versions/web-version/src/game/ball.ts
/**
 * ball.ts -- Unified Ball class for both Classic and Physics modes
 */

import { PhysicsObject } from './physics-object.js';
import { Vector2D } from '../types/index.js';
import {
    COLORS,
    BALL_RADIUS,
    BALL_DEFAULT_VEL,
    MIDDLE_BOARD,
    WIDTH,
    HEIGHT
} from './constants.js';

export type BallMode = 'classic' | 'physics';

export class Ball extends PhysicsObject {
    public radius: number;
    public spin: number;
    public trail: Vector2D[] = [];
    public maxTrail: number = 10;
    public originalPos: Vector2D;
    public originalVel: Vector2D;
    public mode: BallMode;

    constructor(
        x: number = MIDDLE_BOARD[0],
        y: number = MIDDLE_BOARD[1],
        radius: number = BALL_RADIUS,
        color: readonly [number, number, number] = COLORS.WHITE,
        mass: number = 1,
        vel: Vector2D = { x: BALL_DEFAULT_VEL[0], y: BALL_DEFAULT_VEL[1] },
        mode: BallMode = 'physics'
    ) {
        super({ x, y }, vel, { x: 0, y: 0 }, mass, color);
        this.radius = radius;
        this.spin = 0;
        this.originalPos = { x, y };
        this.originalVel = { ...vel };
        this.mode = mode;
    }

    draw(ctx: CanvasRenderingContext2D): void {
        if (this.mode === 'classic') {
            this.drawClassic(ctx);
        } else {
            this.drawPhysics(ctx);
        }
    }

    /**
     * Simple classic mode drawing - just a circle
     */
    private drawClassic(ctx: CanvasRenderingContext2D): void {
        ctx.fillStyle = `rgb(${this.color[0]}, ${this.color[1]}, ${this.color[2]})`;
        ctx.beginPath();
        ctx.arc(this.pos.x, this.pos.y, this.radius, 0, Math.PI * 2);
        ctx.fill();
    }

    /**
     * Physics mode drawing with fire trail, aura, and spin effects
     */
    private drawPhysics(ctx: CanvasRenderingContext2D): void {
        // Fire colors for the trail (from dark to bright)
        const FIRE_COLORS = [
            COLORS.DARK_RED, COLORS.SCARLET, COLORS.ORANGE_RED,
            COLORS.ORANGE, COLORS.GOLD, COLORS.YELLOW, COLORS.WHITE
        ];
        const ALL_FIRE_COLORS = [
            COLORS.DARK_RED, COLORS.SCARLET, COLORS.ORANGE_RED,
            COLORS.ORANGE, COLORS.GOLD, COLORS.YELLOW, COLORS.WHITE,
            COLORS.GOLD, COLORS.YELLOW, COLORS.WHITE
        ];
        const colorCycleLen = ALL_FIRE_COLORS.length;

        const speed = this.speed;
        const maxSpeed = 30;
        const accMagnitude = Math.sqrt(this.acc.x * this.acc.x + this.acc.y * this.acc.y);
        const t = Math.min(accMagnitude / 15, 1);
        const colorIdx = Math.floor(t * (FIRE_COLORS.length - 1));
        const coreColor = FIRE_COLORS[colorIdx];

        // --- 1. Draw fire trail ---
        for (let i = 0; i < this.trail.length; i++) {
            const pos = this.trail[i];
            if (!pos) continue;
            const trailT = i / Math.max(this.trail.length - 1, 1);
            const flicker = speed > 7 ? Math.random() * 32 - 16 : 0;
            const alpha = Math.max(35, 140 - i * 18 + flicker) / 255;
            const radius = Math.max(2, this.radius * (1 - trailT * 0.8));

            const cidx = (colorIdx + i) % colorCycleLen;
            const trailColor = ALL_FIRE_COLORS[cidx];
            if (!trailColor) continue;

            ctx.save();
            ctx.globalAlpha = alpha;
            ctx.fillStyle = `rgb(${trailColor[0]}, ${trailColor[1]}, ${trailColor[2]})`;
            ctx.beginPath();
            ctx.arc(pos.x, pos.y, radius, 0, Math.PI * 2);
            ctx.fill();
            ctx.restore();
        }

        // --- 2. Glowing aura if fast ---
        if (speed > 6) {
            for (let layer = 3; layer > 0; layer--) {
                const auraR = this.radius * (1.2 + 0.25 * layer);
                const auraAlpha = (25 * layer + Math.min(50, speed * 2)) / 255;
                const auraColor = FIRE_COLORS[Math.min(FIRE_COLORS.length - 1, colorIdx + layer)];
                if (!auraColor) continue;

                ctx.save();
                ctx.globalAlpha = auraAlpha;
                ctx.fillStyle = `rgb(${auraColor[0]}, ${auraColor[1]}, ${auraColor[2]})`;
                ctx.beginPath();
                ctx.arc(this.pos.x, this.pos.y, auraR, 0, Math.PI * 2);
                ctx.fill();
                ctx.restore();
            }
        }

        // --- 3. Spin arc effect ---
        if (Math.abs(this.spin) > 4) {
            const direction = this.spin > 0 ? 0 : Math.PI;
            const arcLen = Math.PI + Math.min(Math.PI * 0.8, (speed / maxSpeed) * Math.PI * 0.8);
            const baseThick = 5 + Math.min(18, (speed / maxSpeed) * 22);

            for (let j = 0; j < 5; j++) {
                const arcAlpha = speed > 8
                    ? (60 + (Math.sin(Date.now() / 40 + j) + 1) * 80) / 255
                    : (50 + j * 10) / 255;
                const arcColor = ALL_FIRE_COLORS[(colorIdx + j) % colorCycleLen];
                if (!arcColor) continue;
                const thick = baseThick + j * 2;

                ctx.save();
                ctx.globalAlpha = Math.min(1, arcAlpha);
                ctx.strokeStyle = `rgb(${arcColor[0]}, ${arcColor[1]}, ${arcColor[2]})`;
                ctx.lineWidth = thick;
                ctx.beginPath();
                ctx.arc(this.pos.x, this.pos.y, this.radius, direction, direction + arcLen);
                ctx.stroke();
                ctx.restore();
            }
        }

        // --- 4. Main ball (core) ---
        if (speed > 21 && coreColor) {
            ctx.fillStyle = `rgb(${COLORS.WHITE[0]}, ${COLORS.WHITE[1]}, ${COLORS.WHITE[2]})`;
            ctx.beginPath();
            ctx.arc(this.pos.x, this.pos.y, this.radius, 0, Math.PI * 2);
            ctx.fill();

            ctx.fillStyle = `rgb(${coreColor[0]}, ${coreColor[1]}, ${coreColor[2]})`;
            ctx.beginPath();
            ctx.arc(this.pos.x, this.pos.y, this.radius - 1, 0, Math.PI * 2);
            ctx.fill();
        } else if (coreColor) {
            ctx.fillStyle = `rgb(${coreColor[0]}, ${coreColor[1]}, ${coreColor[2]})`;
            ctx.beginPath();
            ctx.arc(this.pos.x, this.pos.y, this.radius, 0, Math.PI * 2);
            ctx.fill();
        }
    }

    move(): void {
        if (this.mode === 'physics') {
            // Magnus effect for physics mode
            this.vel.y += this.spin * 0.1;
        }
        this.pos.x += this.vel.x;
        this.pos.y += this.vel.y;
    }

    update(): void {
        if (this.mode === 'physics') {
            // Trail only for physics mode
            this.trail.unshift({ x: this.pos.x, y: this.pos.y });
            if (this.trail.length > this.maxTrail) {
                this.trail.pop();
            }
        }
        this.move();
    }

    /**
     * Bounce off walls - used in menu and classic mode
     */
    bounceBox(width: number, height: number): void {
        this.move();
        if (this.pos.x <= 0 || this.pos.x >= width) {
            this.vel.x *= -1;
        }
        if (this.pos.y <= 0 || this.pos.y >= height) {
            this.vel.y *= -1;
        }
    }

    reset(): void {
        this.pos = { ...this.originalPos };
        this.vel = { x: -this.originalVel.x, y: this.mode === 'classic' ? 0 : -this.originalVel.y };
        this.spin = 0;
        this.trail = [];
    }

    get speed(): number {
        return Math.sqrt(this.vel.x * this.vel.x + this.vel.y * this.vel.y);
    }
}

// ----------------------- Legacy Alias for Backwards Compatibility -----------------------
/**
 * @deprecated Use Ball with mode='classic' instead
 */
export class BallClassic extends Ball {
    constructor(
        x: number,
        y: number,
        radius: number,
        color: readonly [number, number, number],
        velX: number,
        velY: number
    ) {
        super(x, y, radius, color, 1, { x: velX, y: velY }, 'classic');
    }
}
